databricks/koalas/frame.py

```diff
@@ -6783,28 +6789,56 @@ def melt(self, id_vars=None, value_vars=None, var_name=None,
         1  b         B          3
         2  c         B          5
         """
+        column_index = self._internal.column_index
+
         if id_vars is None:
             id_vars = []
-        elif isinstance(id_vars, str):
-            id_vars = [(id_vars,)]
-        elif isinstance(id_vars, tuple):
-            if self._internal.column_index_level == 1:
-                id_vars = [idv if isinstance(idv, tuple) else (idv,) for idv in id_vars]
-            else:
-                raise ValueError('id_vars must be a list of tuples when columns are a MultiIndex')
         else:
-            id_vars = [idv if isinstance(idv, tuple) else (idv,) for idv in id_vars]
+            if isinstance(id_vars, str):
+                id_vars = [(id_vars,)]
+            elif isinstance(id_vars, tuple):
+                if self._internal.column_index_level == 1:
+                    id_vars = [idv if isinstance(idv, tuple) else (idv,)
+                               for idv in id_vars]
+                else:
+                    raise ValueError('id_vars must be a list of tuples'
+                                     ' when columns are a MultiIndex')
+            else:
+                id_vars = [idv if isinstance(idv, tuple) else (idv,)
+                           for idv in id_vars]
 
-        column_index = self._internal.column_index
+            raveled_column_index = np.ravel(column_index)
+            missing = [idv for idv in np.ravel(id_vars)
+                       if idv not in raveled_column_index]
```

Comment:
These seem not like a good strategy. E.g., for multi-index columns,

```py
>>> df
   X     Y
   A  B  C
0  a  1  2
1  b  3  4
2  c  5  6
>>> df.melt(value_vars=[('Y', 'A')])
Traceback (most recent call last):
...
pyspark.sql.utils.AnalysisException: "Can't extract value from pairs#68: need struct type but got string;"
```

still throws `AnalysisException`.

Reply:
Wow! I didn’t catch that case.
Thank you. :)